### PR TITLE
Replace systemd-run calls with ssh

### PR DESCRIPTION
**Pull request by gsanchietti**
- Add bash library to be included inside other scripts
- Add nscd-cmd wrapper to execute commands inside the containter
- Add dropbear server inside the container
- Add dropbear client inside the host

NethServer/dev#5544

### CODE REVIEW

**Comment by gsanchietti** [comment]
The PR is ready for review.

File: root/usr/bin/nsdc-cmd [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+      ;;
+  esac
+done
+
+if [ $# -eq 0 ]; then
+    help
+fi
+
+if [ $raw -eq 1 ]; then
+    cmd_raw $@
```
Comment (DavidePrincipi) [inline root/usr/bin/nsdc-cmd]: Use the preserve whitespace syntax

      cmd_raw "$@"

File: root/usr/bin/nsdc-cmd [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+else
+    cmd $@
```
Comment (DavidePrincipi) [inline root/usr/bin/nsdc-cmd]: Same as above

File: root/etc/e-smith/events/actions/nethserver-dc-createadmins [hunk 3]
```diff
@@ -36,11 +36,7 @@ fi
 /etc/e-smith/events/actions/nethserver-dc-user-lock $EVENT "$ADMIN@$DOMAIN"
 
 # add the admin account to Domain Admins group
-expect <<EOF
-spawn -noecho /usr/bin/systemd-run -M nsdc -q -t /usr/bin/samba-tool group addmembers "Domain Admins" "$ADMIN"
-expect "Added members" { exit 0 }
-exit 3
-EOF
+/usr/bin/nsdc-cmd /usr/bin/samba-tool group addmembers \'Domain Admins\' \'$ADMIN\'
```
Comment (DavidePrincipi) [inline root/etc/e-smith/events/actions/nethserver-dc-createadmins:39]: Quotes escaping is no more required once the nscd-cmd fix is appied (see nscd-cmd)

**Review by DavidePrincipi — CHANGES_REQUESTED** [review]
Fix the quotes escaping to preserve whitespace and protect against shell injection

File: root/usr/share/nethserver-dc/nsdc-functions [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+#
+# Functions for nsdc management
+#
+
+# Execute a command inside the container
+#  - redirect all stderr to stdout
+#  - remove dbclient warning output: first 2 lines
+cmd () {
+    LANG=C dbclient -yy -i /var/lib/nethserver/nsdc/nsdc.key root@$(/sbin/e-smith/config getprop nsdc IpAddress) $@ 2>&1 | tail -n+3
```
Comment (DavidePrincipi) [inline root/usr/share/nethserver-dc/nsdc-functions]: Same as above

File: root/usr/share/nethserver-dc/nsdc-functions [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+# Execute a command inside the container
+#   - raw command, no output/input management is performed
+cmd_raw () {
+    LANG=C dbclient -yy -i /var/lib/nethserver/nsdc/nsdc.key root@$(/sbin/e-smith/config getprop nsdc IpAddress) $@
```
Comment (DavidePrincipi) [inline root/usr/share/nethserver-dc/nsdc-functions]: ...and here too

**Comment by DavidePrincipi** [comment]
It's a great idea to replace the bogus systemd-run method! A couple of questions:

- how does authentication work?
- could you add some dev docs?

File: root/usr/bin/nsdc-cmd [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NethServer.  If not, see COPYING.
+#
+
+# Execute a command inside the containter
+
+. /usr/share/nethserver-dc/nsdc-functions
```
Comment (DavidePrincipi) [inline root/usr/bin/nsdc-cmd]: Why not to merge the function definitions here? Do plan to use them directly from other future scripts?

File: root/usr/bin/nsdc-cmd (hunk 1 above)
Comment (gsanchietti) [reply root/usr/bin/nsdc-cmd]: I already tested it, but it doesn't really change much.
Whitespace is already preserved but quotes are not and I didn't find a good way to do it.

But I will add the quotes, for a better implementation.

File: root/etc/e-smith/events/actions/nethserver-dc-createadmins (hunk 3 above)
Comment (gsanchietti) [reply root/etc/e-smith/events/actions/nethserver-dc-createadmins:39]: Quotes are still needed, even after the fix.

File: root/usr/bin/nsdc-cmd (hunk 6 above)
Comment (gsanchietti) [reply root/usr/bin/nsdc-cmd]: Yes, I started creating the functions as a library to be reusable.
But I think we can merge into the command.

**Comment by gsanchietti** [comment]
Fix applied, added documentation.

Ready for new review.

File: root/etc/e-smith/events/actions/nethserver-dc-createadmins (hunk 3 above)
Comment (DavidePrincipi) [reply root/etc/e-smith/events/actions/nethserver-dc-createadmins:39]: Who needs them? It seems quote escaping is required by shell spawned by the dropbear server

```text
[root@vm5 ~]# dbclient  -yy -i /var/lib/nethserver/nsdc/nsdc.key root@192.168.122.55 ls a 'b c'
[...]
[pid  6173] execve("/bin/bash", ["bash", "-c", "ls a b c"], [/* 7 vars */]) = 0
[pid  6173] execve("/usr/bin/ls", ["ls", "a", "b", "c"], [/* 10 vars */]) = 0
```

Compared to

```text
[root@vm5 ~]# dbclient  -yy -i /var/lib/nethserver/nsdc/nsdc.key root@192.168.122.55 ls a \'b c\'
[...]
[pid  6176] execve("/bin/bash", ["bash", "-c", "ls a 'b c'"], [/* 7 vars */]) = 0
[pid  6176] execve("/usr/bin/ls", ["ls", "a", "b c"], [/* 10 vars */]) = 0
```

We must prevent shell injection in general. This issue is not strictly related to dropbear, but we must take it into account, especially where the arguments come from the user's input.

**Comment by DavidePrincipi** [comment]
I submitted an alternative implementation https://github.com/NethServer/nethserver-dc/pull/85. It required a bit of Python code but I prefer it because it does not involve an intermediate shell. As bonus, it does not expose an additional network server.

**Comment by gsanchietti** [comment]
Replaced by #85

**Pull request closed without merging**